### ISSUE TITLE
Implement elchemy-core/XBitwise.elm

### ISSUE THREAD

**Issue by wende** [wende/elchemy#189]
Required to import a lot of binary operation libraries

**Comment by colinbankier** [comment]
I'd be interested in trying to help out with this one - it seems like a pretty straight forward module (maybe?) to get started with. It might take a little while to get familiar with how existing things all work...but I'll start playing around with it.

**Comment by wende** [comment]
@colinbankier That's great!

The only reason why this might be tricky is the usage of many operators, some of which might cause problems when overlapping with Elm's predefined ones.

Whenever you encounter any problem feel free to contact me!

**Comment by colinbankier** [comment]
So most Bitwise functions had a direct elixir/erlang counterpart, except shiftRightZfBy (logical right shift).
It hasn't proven so simple to implement with other existing functions, mostly because erlang doesn't limit the size of integers, so there is no fixed position for a sign bit. I'll keep pondering it. Any thoughts welcome.

**Comment by colinbankier** [comment]
@wende so here is a question. In Javascript (and hence Elm) "The operands of all bitwise operators are converted to signed 32-bit integers in two's complement format." from https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Bitwise_Operators

Erlang, however, doesn't limit the size on integers.

Hence a bitwise shift in Elm "rolls over" the 32 bits:
```
> Bitwise.shiftLeftBy 32 1
1 : Int
```

But in erlang/elixir, the number keeps increasing
```
iex(141)> Bitwise.bsl(1, 32)        
4294967296
```

If we convert everything to signed 32-bit integers, we can remain consistent with Elm (and implementing shiftRightZfBy becomes easy), but not with elixir. I assume the goal is consistency with Elm - to reuse existing libraries. Thoughts?

**Comment by wende** [comment]
@colinbankier Oooh. Tough one. Although Elixir's usability seems like a little wider I think we should definitely aim for consistency with the original core library.
We don't want somebody to use Bitwise in his library both in Elm and Elchemy an experience different behaviours.

**Comment by wende** [comment]
Closed by  wende/elchemy-core#6
